File: src/milestones.ts
import * as p from "@clack/prompts";
import { preflight } from "./preflight.ts";
import { loadConfig } from "./config.ts";
import {
  listMilestones,
  createMilestone,
  updateMilestone,
} from "./github.ts";

function addDays(date: Date, days: number): Date {
  const result = new Date(date);
  result.setDate(result.getDate() + days);
  return result;
}

function formatDate(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

export async function milestones(): Promise<void> {
  p.intro("gh-setup milestones");

  const preflightResult = await preflight();
  if (!preflightResult) {
    p.outro("Cancelled.");
    process.exitCode = 1;
    return;
  }

  const { repo } = preflightResult;
  p.log.info(`Repository: ${repo}`);

  // Load config or prompt interactively
  const config = loadConfig();
  let startDate: string;
  let weeks: number;

  if (config?.milestones) {
    startDate = config.milestones.startDate;
    weeks = config.milestones.weeks;
    p.log.info(`Config: startDate=${startDate}, weeks=${weeks}`);
  } else {
    const startDateInput = await p.text({
      message: "Start date (first Sunday, YYYY-MM-DD):",
      placeholder: "2026-01-04",
      validate: (v) =>
        v && /^\d{4}-\d{2}-\d{2}$/.test(v)
          ? undefined
          : "Format: YYYY-MM-DD",
    });
    if (p.isCancel(startDateInput)) {
      p.outro("Cancelled.");
      return;
    }

    const weeksInput = await p.text({
      message: "Number of weeks:",
      placeholder: "52",
      validate: (v) => {
        const n = Number(v);
        return Number.isInteger(n) && n > 0
          ? undefined
          : "Enter a positive integer";
      },
    });
    if (p.isCancel(weeksInput)) {
      p.outro("Cancelled.");
      return;
    }

    startDate = startDateInput;
    weeks = Number(weeksInput);
  }

  // Fetch existing milestones
  const s = p.spinner();
  s.start("Fetching existing milestones...");

  let existingMap: Map<string, number>;
  try {
    const existing = await listMilestones(repo);
    existingMap = new Map();
    for (const m of existing) {
      if (m.due_on) {
        const dateKey = m.due_on.slice(0, 10);
        if (existingMap.has(dateKey)) {
          p.log.warn(
            `Duplicate due_on date ${dateKey}: milestone #${m.number} ("${m.title}") overwrites #${existingMap.get(dateKey)}`,
          );
        }
        existingMap.set(dateKey, m.number);
      }
    }
    s.stop(`Found ${existing.length} existing milestones.`);
  } catch (e) {
    s.stop("Failed to fetch milestones.");
    p.log.error(e instanceof Error ? e.message : String(e));
    process.exitCode = 1;
    return;
  }

  // Confirmation
  const confirmed = await p.confirm({
    message: `Create/update ${weeks} weekly milestones starting from ${startDate}?`,
  });
  if (p.isCancel(confirmed) || !confirmed) {
    p.outro("Cancelled.");
    return;
  }

  // Create/update milestones
  let created = 0;
  let updated = 0;
  let failed = 0;
  const failures: { title: string; error: string }[] = [];

  s.start("Creating milestones...");

  for (let i = 0; i < weeks; i++) {
    const weekStart = addDays(new Date(startDate + "T00:00:00"), i * 7);
    const weekEnd = addDays(weekStart, 6); // Saturday
    const weekNum = i + 1;
    const endDateStr = formatDate(weekEnd);
    const title = `Week ${weekNum}: ${endDateStr}`;
    const description = `Period: ${formatDate(weekStart)} - ${endDateStr}`;

    const dueTime = config?.milestones?.dueTime ?? "14:59:59Z";
    const dueOn = `${endDateStr}T${dueTime}`;

    const existingNumber = existingMap.get(endDateStr);
    s.message(`${existingNumber ? "Updating" : "Creating"}: ${title}`);

    try {
      if (existingNumber) {
        await updateMilestone(repo, existingNumber, title, description);
        updated++;
      } else {
        await createMilestone(repo, title, description, dueOn);
        created++;
      }
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      failures.push({ title, error: msg });
      failed++;
    }
  }

  s.stop("Done.");

  for (const f of failures) {
    p.log.error(`  ${f.title}: ${f.error}`);
  }

  p.log.info(`Created: ${created} / Updated: ${updated} / Failed: ${failed}`);
  if (failed > 0) {
    process.exitCode = 1;
  }
  p.outro("Milestones complete!");
}

File: cmd/milestones.go
package cmd

import (
	"fmt"
	"strconv"
	"time"

	"github.com/spf13/cobra"
	"github.com/ynamiki-mediee/gh-setup/internal/config"
	"github.com/ynamiki-mediee/gh-setup/internal/gh"
	"github.com/ynamiki-mediee/gh-setup/internal/milestone"
	"github.com/ynamiki-mediee/gh-setup/internal/prompt"
)

var milestonesCmd = &cobra.Command{
	Use:   "milestones",
	Short: "Create or update weekly milestones",
	RunE:  runMilestones,
}

func init() {
	rootCmd.AddCommand(milestonesCmd)
}

var timezoneOptions = []struct{ label, value string }{
	{"Asia/Tokyo (JST, UTC+9)", "Asia/Tokyo"},
	{"Asia/Shanghai (CST, UTC+8)", "Asia/Shanghai"},
	{"Asia/Kolkata (IST, UTC+5:30)", "Asia/Kolkata"},
	{"Europe/Berlin (CET, UTC+1)", "Europe/Berlin"},
	{"Europe/London (GMT, UTC+0)", "Europe/London"},
	{"America/New_York (EST, UTC-5)", "America/New_York"},
	{"America/Chicago (CST, UTC-6)", "America/Chicago"},
	{"America/Los_Angeles (PST, UTC-8)", "America/Los_Angeles"},
	{"UTC", "UTC"},
}

func runMilestones(cmd *cobra.Command, args []string) error {
	// Step 1: Detect and confirm repo
	detected, host, err := gh.DetectRepo()
	if err != nil {
		return err
	}

	repo, err := prompt.ConfirmRepo(detected)
	if err != nil {
		if handleCancel(err) {
			return nil
		}
		return err
	}

	// Step 2-3: Get milestone parameters from config or interactive prompts
	var startDate string
	var weeks int
	var timezone string

	cfg, err := config.LoadConfig()
	if err != nil {
		return err
	}

	if cfg != nil && cfg.Milestones != nil {
		startDate = cfg.Milestones.StartDate
		weeks = cfg.Milestones.Weeks
		timezone = cfg.Milestones.Timezone
		if timezone == "" {
			timezone = "UTC"
		}
	} else {
		// Interactive prompts
		startDate, err = prompt.TextInput("Start date (YYYY-MM-DD):", "", milestone.NextMonday(time.Now()))
		if err != nil {
			if handleCancel(err) {
				return nil
			}
			return err
		}

		weeksUntilEnd, err := milestone.WeeksUntilEndOfYear(startDate)
		if err != nil {
			return err
		}
		weeksDefault := strconv.Itoa(weeksUntilEnd)
		weeksStr, err := prompt.TextInput("Number of weeks:", "", weeksDefault)
		if err != nil {
			if handleCancel(err) {
				return nil
			}
			return err
		}
		weeks, err = strconv.Atoi(weeksStr)
		if err != nil {
			return fmt.Errorf("invalid number of weeks %q: %w", weeksStr, err)
		}

		// Build timezone select
		labels := make([]string, len(timezoneOptions))
		for i, o := range timezoneOptions {
			labels[i] = o.label
		}

		// Auto-detect system timezone
		systemTZ := time.Now().Location().String()
		defaultLabel := "UTC" // last option label
		for _, o := range timezoneOptions {
			if o.value == systemTZ {
				defaultLabel = o.label
				break
			}
		}

		selected, err := prompt.SelectWithDefault("Timezone for due dates:", labels, defaultLabel)
		if err != nil {
			if handleCancel(err) {
				return nil
			}
			return err
		}

		// Map selected label back to value
		for _, o := range timezoneOptions {
			if o.label == selected {
				timezone = o.value
				break
			}
		}
	}

	// Step 4: Create client and fetch existing milestones
	client, err := gh.NewClient(host)
	if err != nil {
		return err
	}

	fmt.Println("Fetching existing milestones...")
	existing, err := client.ListMilestones(repo)
	if err != nil {
		return err
	}

	// Step 5: Build existing map: title → milestone number
	existingMap := make(map[string]int)
	for _, m := range existing {
		existingMap[m.Title] = m.Number
	}

	// Step 6: Confirm
	ok, err := prompt.Confirm(fmt.Sprintf("Create/update %d weekly milestones starting from %s?", weeks, startDate))
	if err != nil {
		if handleCancel(err) {
			return nil
		}
		return err
	}
	if !ok {
		fmt.Println("Cancelled.")
		return nil
	}

	// Step 7: Loop and create/update milestones
	baseDate, err := time.Parse("2006-01-02", startDate)
	if err != nil {
		return fmt.Errorf("invalid start date %q: %w", startDate, err)
	}

	var created, updated, failed int
	var failures []string

	for i := 0; i < weeks; i++ {
		weekStart := baseDate.AddDate(0, 0, i*7)
		weekEnd := weekStart.AddDate(0, 0, 6)

		weekNum := milestone.ISOWeek(weekStart)
		startDateStr := weekStart.Format("2006-01-02")
		endDateStr := weekEnd.Format("2006-01-02")

		title := fmt.Sprintf("Week %d: %s", weekNum, endDateStr)
		description := fmt.Sprintf("Period: %s - %s", startDateStr, endDateStr)

		dueOn, err := milestone.ToUtcDueOn(endDateStr, timezone)
		if err != nil {
			msg := fmt.Sprintf("Failed to compute due_on for %s: %v", endDateStr, err)
			fmt.Println(msg)
			failures = append(failures, msg)
			failed++
			continue
		}

		if num, exists := existingMap[title]; exists {
			// Update existing milestone
			fmt.Printf("Updating: %s...\n", title)
			if err := client.UpdateMilestone(repo, num, title, description, dueOn); err != nil {
				fmt.Printf("✗ %s: %v\n", title, err)
				failures = append(failures, fmt.Sprintf("Failed to update milestone %q: %v", title, err))
				failed++
			} else {
				fmt.Printf("✓ %s\n", title)
				updated++
			}
		} else {
			// Create new milestone
			fmt.Printf("Creating: %s...\n", title)
			if err := client.CreateMilestone(repo, title, description, dueOn); err != nil {
				fmt.Printf("✗ %s: %v\n", title, err)
				failures = append(failures, fmt.Sprintf("Failed to create milestone %q: %v", title, err))
				failed++
			} else {
				fmt.Printf("✓ %s\n", title)
				created++
			}
		}
	}

	// Step 8: Print summary
	if len(failures) > 0 {
		fmt.Println("\nFailures:")
		for _, f := range failures {
			fmt.Printf("  - %s\n", f)
		}
	}

	fmt.Printf("Created: %d / Updated: %d / Failed: %d\n", created, updated, failed)

	if failed > 0 {
		return fmt.Errorf("%d milestone operation(s) failed", failed)
	}

	return nil
}
